docs/src/main/paradox/user/lang/go/src/eventsourced.go

```diff
@@ -0,0 +1,565 @@
+	arg1 := method.Type.In(2)
+	ptr := false
+	for arg1.Kind() == reflect.Ptr {
+		ptr = true
+		arg1 = arg1.Elem()
+	}
+	var msg proto.Message
+	if ptr {
+		msg = reflect.New(arg1).Interface().(proto.Message)
+	} else {
+		msg = reflect.Zero(arg1).Interface().(proto.Message)
+	}
+	if err := proto.Unmarshal(cmd.GetPayload().GetValue(), msg); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal: %w", err)
+	}
+	inputs[2] = reflect.ValueOf(msg)
+	return inputs, nil
+}
+
+func (esh *EventSourcedHandler) methodToCall(cmd *protocol.Command) (reflect.Method, error) {
+	entityContext := esh.contexts[cmd.GetEntityId()]
+	cacheKey := entityContext.ServiceName() + cmd.Name
+	method, hit := esh.cmdMethodCache[cacheKey]
+	// as measured this cache saves us about 75% of a call
+	// to be prepared with 4.4µs vs. 17.6µs where a typical
+	// call by reflection like GetCart() with Func.Call()
+	// takes ~10µs and to get return values processed somewhere 0.7µs.
+	if !hit {
+		entityValue := reflect.ValueOf(entityContext.EntityInstance.Instance)
+		// entities implement the proxied grpc service
+		// we try to find the method we're called by name with the
+		// received command.
+		methodByName := entityValue.MethodByName(cmd.Name)
+		if !methodByName.IsValid() {
+			entity := esh.entities[entityContext.ServiceName()]
+			return reflect.Method{}, fmt.Errorf("no method named: %s found for: %v", cmd.Name, entity)
+		}
+		// gRPC services are unary rpc methods, always.
+		// They have one message in and one message out.
+		if err := checkUnary(methodByName); err != nil {
+			return reflect.Method{}, err
+		}
+		// The first argument in the gRPC implementation
+		// is always a context.Context.
+		methodArg0Type := methodByName.Type().In(0)
+		contextType := reflect.TypeOf(context.Background())
+		if !contextType.Implements(methodArg0Type) {
+			return reflect.Method{}, fmt.Errorf(
+				"first argument for method: %s is not of type: %s",
+				methodByName.String(), contextType.Name(),
+			)
+		}
+		// we'll find one for sure as we found one on the entityValue
+		method, _ = reflect.TypeOf(entityContext.EntityInstance.Instance).MethodByName(cmd.Name)
+		esh.cmdMethodCache[cacheKey] = method
+	}
+	return method, nil
+}
+
+func (*EventSourcedHandler) handleSnapshots(entityContext *EntityInstanceContext) (*any.Any, error) {
+	if !entityContext.EntityInstance.shouldSnapshot() {
+		return nil, nil
+	}
+	if snapshotter, canSnapshot := entityContext.EntityInstance.Instance.(Snapshotter); canSnapshot {
+		snap, err := snapshotter.Snapshot()
+		if err != nil {
+			return nil, fmt.Errorf("getting a snapshot has failed: %v. %w", err, ErrFailure)
+		}
+		// TODO: we expect a proto.Message but should support other formats
+		snapshot, err := marshalAny(snap)
+		if err != nil {
+			return nil, err
+		}
+		entityContext.EntityInstance.resetSnapshotEvery()
+		return snapshot, nil
+	} else {
+		// TODO: every entity should implement snapshotting, right?
```

Comment:
See: 
https://github.com/cloudstateio/cloudstate/blob/master/java-support/src/main/scala/io/cloudstate/javasupport/impl/eventsourced/EventSourcedImpl.scala#L233

and

https://github.com/cloudstateio/cloudstate/blob/master/java-support/src/main/scala/io/cloudstate/javasupport/impl/eventsourced/EventSourcedImpl.scala#L166

Reply:
I will address it in the go-support repo.